tests/test_other.py

```diff
@@ -437,43 +437,55 @@ def test_emcc_5(self):
         assert test(open('a.out.js').read()), text
 
-  def test_emcc_6(self):
-    for compiler in [EMCC, EMXX]:
-      # Compiling two source files into a final JS.
-      for args, target in [([], 'a.out.js'), (['-o', 'combined.js'], 'combined.js')]:
-        self.clear()
-        run_process([PYTHON, compiler, path_from_root('tests', 'twopart_main.cpp'), path_from_root('tests', 'twopart_side.cpp')] + args)
-        self.assertContained('side got: hello from main, over', run_js(target))
+  def test_multiple_sources(self):
+    # Compiling two sources at a time should work.
+    cmd = [PYTHON, EMCC, '-c', path_from_root('tests', 'twopart_main.cpp'), path_from_root('tests', 'twopart_side.cpp')]
+    run_process(cmd)
 
-        # Compiling two files with -c will generate separate .bc files
-        self.clear()
-        expect_error = '-o' in args # specifying -o and -c is an error
-        cmd = [PYTHON, compiler, path_from_root('tests', 'twopart_main.cpp'), path_from_root('tests', 'twopart_side.cpp'), '-c'] + args
-        if expect_error:
-          err = self.expect_fail(cmd)
-          self.assertContained('fatal error', err)
-          continue
+    # Object files should be generated by default in the current working
+    # directory, and not alongside the sources.
+    self.assertExists('twopart_main.o')
+    self.assertExists('twopart_side.o')
+    self.assertNotExists(path_from_root('tests', 'twopart_main.o'))
+    self.assertNotExists(path_from_root('tests', 'twopart_side.o'))
 
-        run_process(cmd)
-        try_delete(target)
-
-        # Compiling one of them alone is expected to fail
-        run_process([PYTHON, compiler, 'twopart_main.o', '-O1', '-g', '-s', 'WARN_ON_UNDEFINED_SYMBOLS=0'] + args)
-        self.assertContained('missing function', run_js(target, stderr=STDOUT, assert_returncode=None))
-        try_delete(target)
-
-        # Combining those object files into js should work
-        run_process([PYTHON, compiler, 'twopart_main.o', 'twopart_side.o'] + args)
-        self.assertContained('side got: hello from main, over', run_js(target))
-
-        # Combining object files into another object should also work
-        try_delete(target)
-        run_process([PYTHON, compiler, 'twopart_main.o', 'twopart_side.o', '-o', 'combined.o'] + args)
-        syms = Building.llvm_nm('combined.o')
-        assert len(syms.defs) in (2, 3) and 'main' in syms.defs, 'Should be two functions (and in the wasm backend, also __original_main)'
-        run_process([PYTHON, compiler, 'combined.o', '-o', 'combined.o.js'])
-        self.assertExists('combined.o.js')
-        self.assertContained('side got: hello from main, over', run_js('combined.o.js'))
+    # But it is an error if '-o' is also specified.
+    self.clear()
+    err = self.expect_fail(cmd + ['-o', 'out.o'])
+
+    self.assertContained('cannot specify -o with -c/S and multiple source files', err)
+    self.assertNotExists('twopart_main.o')
```

Comment:
It's hard to tell due to the cleanup+restructuring of the tests, but am I correct that this is the real functional change?
Specifically, error earlier, to avoid creating object files we can't do anything with in the first place.

That's a good change but it wasn't clear.

Reply:
Yes we had a case where were leaking object files alongside the source files.

There is more extensive fix I could do, but moving the check earlier before we do any compilation seems simpler for now, as well as strictly better.